language-plutus-core/src/Language/PlutusCore/Check/Normal.hs

```diff
@@ -2,105 +2,56 @@
+                                        , isNormalType
                                         ) where
 
+import           Control.Lens
 import           Control.Monad.Except
 
-import           Data.Functor.Foldable
-import           Data.Functor.Foldable.Monadic
-
 import           Language.PlutusCore.Error
 import           Language.PlutusCore.Name
 import           Language.PlutusCore.Type
 import           PlutusPrelude
 
--- | Ensure that all terms and types are well-formed accoring to Fig. 2
+-- | Ensure that all types in the 'Program' are normalized.
 checkProgram :: (AsNormalizationError e TyName Name a, MonadError e m) => Program TyName Name a -> m ()
-checkProgram p = void $ throwingEither _NormalizationError $ preCheck p
+checkProgram (Program _ _ t) = checkTerm t
 
--- | Ensure that all terms and types are well-formed accoring to Fig. 2
+-- | Ensure that all types in the 'Term' are normalized.
 checkTerm :: (AsNormalizationError e TyName Name a, MonadError e m) => Term TyName Name a -> m ()
-checkTerm p = void $ throwingEither _NormalizationError $ checkTerm p
-
-check :: Program tyname name a -> Maybe (NormalizationError tyname name a)
-check = go . preCheck where
-    go Right{}  = Nothing
-    go (Left x) = Just x
-
--- | Ensure that all terms and types are well-formed accoring to Fig. 2
-preCheck :: Program tyname name a -> Either (NormalizationError tyname name a) (Program tyname name a)
-preCheck (Program l v t) = Program l v <$> checkT t
-
--- this basically ensures all type instatiations, etc. occur only with type *values*
-checkT :: Term tyname name a -> Either (NormalizationError tyname name a) (Term tyname name a)
-checkT (Error l ty)           = Error l <$> typeValue ty
-checkT (TyInst l t ty)        = TyInst l <$> checkT t <*> typeValue ty
-checkT (IWrap l pat arg term) = IWrap l <$> typeValue pat <*> typeValue arg <*> checkT term
-checkT (Unwrap l t)           = Unwrap l <$> checkT t
-checkT (LamAbs l n ty t)      = LamAbs l n <$> typeValue ty <*> checkT t
-checkT (Apply l t t')         = Apply l <$> checkT t <*> checkT t'
-checkT (TyAbs l tn k t)       = TyAbs l tn k <$> termValue t
-checkT t@Var{}                = pure t
-checkT t@Constant{}           = pure t
-checkT t@Builtin{}            = pure t
-
-isTermValue :: Term tyname name a -> Bool
-isTermValue = isRight . termValue
+checkTerm t = void $ throwingEither _NormalizationError $ check t
 
--- ensure a term is a value
-termValue :: Term tyname name a -> Either (NormalizationError tyname name a) (Term tyname name a)
-termValue (LamAbs l n ty t)      = LamAbs l n ty <$> checkT t
-termValue (IWrap l pat arg term) = IWrap l pat arg <$> termValue term
-termValue (TyAbs l tn k t)       = TyAbs l tn k <$> termValue t
-termValue t                      = builtinValue t
+check :: Term tyname name a -> Either (NormalizationError tyname name a) (Term tyname name a)
+check = transformMOf termSubterms (termSubtypes normalType)
```

Comment:
What was wrong with the previous version? That style is clearer to me.

Reply:
It's the same as we did for `normalizeTypesInM`. If we think that was clearer then I don't think this is any different?

I'm not massively attached to it, but I do like effectively saying "look at all the types attached to all transitive subterms and check that they are normal", since that's pretty much the statement of what we want to do.

Reply:
As with `normalizeTypesInM` I like this version more than the previous one, however I think the previous one is weird, because it returns the same term back and so we're building on a wrong approach and should fix the approach and build on the right one.

Reply:
No longer using the traversals.